### DISCUSSION TITLE
截图功能碰到的问题及讨论

### DISCUSSION THREAD

**Discussion by FickXu** [yume-chan/ya-webadb#484] [Q&A]
最初demo截图功能正常，但使用一段时间后出现了一个错误：
![image](https://user-images.githubusercontent.com/7928985/216261586-b92e5c62-15cc-490b-bec0-80644a7fc652.png)
这是websocket的返回：
![image](https://user-images.githubusercontent.com/7928985/216262001-50bfee34-0d01-48d8-95cb-d4b372bc9850.png)
服务器推送的流数据持续到这个时候：
![image](https://user-images.githubusercontent.com/7928985/216262232-4bd140c6-037e-431d-8e1f-19047dd8a25e.png)
持续了20多秒，每秒大概有20多个10.2kb的流数据，总共大概有4M，然后界面弹出错误提示。
我推测是图片太大，导致流处理的过程发生了异常。请问我需要怎么定位处理这个问题？
我已经尝试在依赖包源码中打印日志来调试，但里面逻辑太多，ts相关的技术我并不熟悉，所以进度很慢。希望你能提供一些帮助。

**Comment by yume-chan** [comment]
I don't know, there are so many possibilities.

The WebSocket spec defines that if the send buffer is full (when network is slow or the other end consumes data too slowly), further calls to `send` can close the connection. It depends on the WebSocket server library you are using. For example, with `ws` package for Node.js, the send buffer can grow infinitely, so this won't happen.

There are also some reports about Chrome missing WebSocket messages when large amount of data is sent, for example https://github.com/whatwg/websockets/issues/33, https://github.com/websockets/ws/issues/2042 and https://crbug.com/1214238. However, none of them can be confirmed. To debug this, capture and analyze raw network traffic using software like WireShark.

You might start from something easy. You said the total amount of data received is around 4 MB. Android framebuffer command always uses RGBA8888 format, so the expected image size should be `width x height x 4 bytes`. Does the size match?

Also, you can compare the packets when it's working vs not working, or check the ADB packet log when connected over USB vs wirelessly. The characteristics of framebuffer packets are very distinctive, every fourth byte is `0xff`, because screen doesn't have alpha channel.
